### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [361]:
# Dependencies and Setup
import pandas as pd
import numpy as np

# File to Load
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file_to_load)
# purchase_data.head()

# purchase_data.columns

## Player Count

* Display the total number of players


In [362]:
# Calculating total number of unique SN's
total_player_count = len(purchase_data["SN"].unique())

# Createing DataFrame to display total unique players
total_players_df = pd.DataFrame({"Total Players": [total_player_count]})
total_players_df

Total Players
0            576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [363]:
# Calculating total unique items
number_unique_items = len(purchase_data["Item ID"].unique())
# number_unique_items

# Calculating average price per item
average_price = round((purchase_data["Price"]).mean(),2)
# average_price

# Calculating number of purchases
number_purchases = len(purchase_data["Purchase ID"])
# number_purchases

# Calculating the total revenue
revenue_total = purchase_data["Price"].sum()
# revenue_total

# Creating DataFrame for all above calculations concerning unique items, avererage price per item, number of purchases, and total revenue
purchasing_analysis_df = pd.DataFrame(
    {"Number_of_Unique_Items": [number_unique_items],
     "Average_Price": [average_price],
     "Number_of_Purchase": [number_purchases],
     "Total_Revenue": [revenue_total]})
# purchasing_analysis_df

# Formatting the DataFrame
purchasing_analysis_df["Average_Price"] = purchasing_analysis_df["Average_Price"].map("${:.2f}".format)
purchasing_analysis_df["Total_Revenue"] = purchasing_analysis_df["Total_Revenue"].map("${:.2f}".format)
purchasing_analysis_df

Number_of_Unique_Items Average_Price  Number_of_Purchase Total_Revenue
0                     183         $3.05                 780      $2379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [364]:
# Calculating non-duplicate gender counts and percentages
non_duplicate = purchase_data.drop_duplicates("SN")
total_gender_count = non_duplicate["Gender"].value_counts().sum()
total_gender_percent = non_duplicate["Gender"].value_counts()/total_gender_count

# Creating variables to put into DataFrame
gender_percent_list = total_gender_percent.values
gender_count_list = non_duplicate["Gender"].value_counts().values
gender_list = non_duplicate["Gender"].value_counts().index

# Creating DataFrame for Gender Count and Percentages of Genders
summDf = pd.DataFrame()
summDf["Genders"] = gender_list
summDf["Total_Count"] = gender_count_list
summDf["Percentage_of_Players"] = gender_percent_list * 100
summDf

Genders  Total_Count  Percentage_of_Players
0                   Male          484              84.027778
1                 Female           81              14.062500
2  Other / Non-Disclosed           11               1.909722

In [365]:
# Creating a sorted variable of the gender list above for calculating average total purchase per person below
sortedDF= summDf.sort_values("Genders").reset_index(drop=True)
# sortedDF


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [366]:
# Calculating purchase counts by Gender
purchase_count = purchase_data["Gender"].value_counts()
gender_purchase_count = pd.DataFrame(purchase_count)
gender_purchase_count = gender_purchase_count.rename(columns={"Gender": "Purchase_Count"})
purchase_count_list = list(gender_purchase_count["Purchase_Count"])
gender_count_list = list(gender_purchase_count.index)
# print(purchase_count_list)
# print(gender_count_list)

In [367]:
# Calculating Gender purchase averages
gender_grouped = purchase_data.groupby(["Gender"]).mean()
ave_purchase_price = gender_grouped.Price.values
# ave_purchase_price

In [368]:
# Calculating total purchase value by Gender
gender_grouped = purchase_data.groupby(["Gender"]).sum()
total_purchase_value = gender_grouped.Price.values

In [369]:
# Creating Gender Purchase Count list
gender_purchase_count = list(purchase_data.groupby(["Gender"]).count().Price.values)

In [370]:
# Creating the Gender List
gender_list = list(purchase_data.groupby(["Gender"]).sum().index)

In [371]:
# Calculating average total purchase by Gender
ave_total_purchase = gender_grouped.Price.values/sortedDF.Total_Count

In [372]:
# Creating a dictionary to put into a DataFrame
gender_dict = {"Gender": gender_list,
               "Purchase_Count": gender_purchase_count,
               "Average_Purchase_Price": ave_purchase_price,
               "Total_Purchase_Value": total_purchase_value,
               "Average_Total_Purchase_Per_Person": ave_total_purchase}
# gender_dict

# Created a DataFrame using the dictionary above to show the total purchase count by gender, average purchase price by gender, total purchase value by gender, and average total purchase per person by gender
purchasing_analysis = pd.DataFrame(gender_dict)
# purchasing_analysis

# Formatting the DataFrame
purchasing_analysis["Average_Purchase_Price"] = purchasing_analysis["Average_Purchase_Price"].map("${:.2f}".format)
purchasing_analysis["Total_Purchase_Value"] = purchasing_analysis["Total_Purchase_Value"].map("${:.2f}".format)
purchasing_analysis["Average_Total_Purchase_Per_Person"] = purchasing_analysis["Average_Total_Purchase_Per_Person"].map("${:.2f}".format)
purchasing_analysis

Gender  Purchase_Count Average_Purchase_Price  \
0                 Female             113                  $3.20   
1                   Male             652                  $3.02   
2  Other / Non-Disclosed              15                  $3.35   

  Total_Purchase_Value Average_Total_Purchase_Per_Person  
0              $361.94                             $4.47  
1             $1967.64                             $4.07  
2               $50.19                             $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [373]:
# Creating bins and groups to summarize age data into
bins = [0, 9.9, 14.9, 19.9, 24.9, 29.9, 34.9, 39.9, 150]
group = ("<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+")

# Putting the age into age buckets
summary_views = pd.cut(non_duplicate["Age"],bins,labels=group)
# summary_views

# Creating new DataFrame with age groups
non_duplicate["Total Count"] = pd.cut(non_duplicate["Age"],bins,labels=group)
# non_duplicate.head()

# Count of Totals by Age Group
age_count_df = non_duplicate.groupby("Total Count")["Item Name"]
# age_count_df.count()

# Percentage of Totals by Age Group
age_percentage_df = non_duplicate["Total Count"].value_counts()/total_gender_count
# age_percentage_df

# Creating the actual DataFrame
age_buckets = {"Total Count": (age_count_df.count()),
               "Percentage_of_Players": (round(age_percentage_df * 100,2)) }
# age_buckets

# Printing the DataFrame to show the count of each age group and percentage of each age group
age_bucket_analysis = pd.DataFrame(age_buckets).sort_index()
age_bucket_analysis


//anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  # Remove the CWD from sys.path while we load stuff.


Total Count  Percentage_of_Players
<10             17                   2.95
10-14           22                   3.82
15-19          107                  18.58
20-24          258                  44.79
25-29           77                  13.37
30-34           52                   9.03
35-39           31                   5.38
40+             12                   2.08

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [374]:
# Recreating the bins and groups to calculate the new analyses
bins = [0, 9.9, 14.9, 19.9, 24.9, 29.9, 34.9, 39.9, 150]
group = ("<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+")

# Putting the age into age buckets
summary_views1 = pd.cut(purchase_data["Age"],bins,labels=group)
# summary_views1

# Creating a DataFrame with age groups
purchase_data["Total Count"] = pd.cut(purchase_data["Age"],bins,labels=group)
# purchase_data.head()

# Count of Totals by Age Group
age_count_df1 = purchase_data.groupby("Total Count")["Item Name"]
# age_count_df1.count()


# Average Purchase Price
age_count_grouped = purchase_data.groupby(["Total Count"]).mean()

ave_age_purchase_price = age_count_grouped.Price.values
# ave_age_purchase_price


# Total Purchase Value
age_grouped = purchase_data.groupby(["Total Count"]).sum()
total_purchase_value_age = age_grouped.Price.values
# total_purchase_value_age


# Average Purchase Total per Person
unique_age_count = non_duplicate.groupby("Total Count")["Item Name"]
# unique_age_count.count()

# Calculating the average total purchase per age group
ave_total_purchase_age = (total_purchase_value_age/(unique_age_count.count()))
# ave_total_purchase_age

# Creating a Dictionary
age_demo_buckets = {"Purchase_Count": age_count_df1.count(),
                    "Average_Purchase_Price": list(ave_age_purchase_price),
                    "Total_Purchase_Value": list(total_purchase_value_age),
                    "Average_Purchase_Total_per_Person": list(ave_total_purchase_age)}
# age_demo_buckets

# Printing the DataFrame to show the purchase count per age group, the average purchase 
# price per age group, the total purchase vale per age group, and the average purchase total per person 
age_demo_bucket_analysis = pd.DataFrame(age_demo_buckets)
# age_demo_bucket_analysis

# Formatting the DataFrame
age_demo_bucket_analysis["Average_Purchase_Price"] = age_demo_bucket_analysis["Average_Purchase_Price"].map("${:.2f}".format)
age_demo_bucket_analysis["Total_Purchase_Value"] = age_demo_bucket_analysis["Total_Purchase_Value"].map("${:.2f}".format)
age_demo_bucket_analysis["Average_Purchase_Total_per_Person"] = age_demo_bucket_analysis["Average_Purchase_Total_per_Person"].map("${:.2f}".format)
age_demo_bucket_analysis


Purchase_Count Average_Purchase_Price Total_Purchase_Value  \
Total Count                                                               
<10                      23                  $3.35               $77.13   
10-14                    28                  $2.96               $82.78   
15-19                   136                  $3.04              $412.89   
20-24                   365                  $3.05             $1114.06   
25-29                   101                  $2.90              $293.00   
30-34                    73                  $2.93              $214.00   
35-39                    41                  $3.60              $147.67   
40+                      13                  $2.94               $38.24   

            Average_Purchase_Total_per_Person  
Total Count                                    
<10                                     $4.54  
10-14                                   $3.76  
15-19                                   $3.86  
20-24                                   $4.32  
25-29                                   $3.81  
30-34                                   $4.12  
35-39                                   $4.76  
40+                                     $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [375]:
# Create variables to put into DataFrame
SN_count = purchase_data.groupby("SN").count()["Price"]
SN_average_price = purchase_data.groupby("SN").mean()["Price"]
SN_total_value = purchase_data.groupby("SN").sum()["Price"]

# Creating a DataFrame
top_spenders_df = pd.DataFrame({"Purchase_Count": SN_count,
                                "Average_Purchase_Price": SN_average_price,
                                "Total_Purchase_Value": SN_total_value})

# Sort DataFrame for top 5 Spenders
top_five_spenders = top_spenders_df.sort_values("Total_Purchase_Value", ascending=False)
# top_five_spenders.head(5)

# Formatting the DataFrame
top_five_spenders["Average_Purchase_Price"] = top_five_spenders["Average_Purchase_Price"].map("${:.2f}".format)
top_five_spenders["Total_Purchase_Value"] = top_five_spenders["Total_Purchase_Value"].map("${:.2f}".format)
top_five_spenders.head(5)

Purchase_Count Average_Purchase_Price Total_Purchase_Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [376]:
# Create variables to put into DataFrame
item_count = purchase_data.groupby(["Item ID", "Item Name"]).count()["Price"]
average_item_purchase = purchase_data.groupby(["Item ID", "Item Name"]).mean()["Price"]
total_item_purchase = purchase_data.groupby(["Item ID", "Item Name"]).sum()["Price"]

# Creating a DataFrame
popular_items_df = pd.DataFrame({"Purchase_Count": item_count,
                                 "Item_Price": average_item_purchase,
                                 "Total_Purchase_Value": total_item_purchase})

# Sort DataFrame for top 5 Most Popular Items
most_popular_items_df = popular_items_df.sort_values("Purchase_Count", ascending=False)
most_popular_items_df.head(5)



Purchase_Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
108     Extraction, Quickblade Of Trembling Hands                  9   
82      Nirvana                                                    9   
19      Pursuit, Cudgel of Necromancy                              8   

                                                      Item_Price  \
Item ID Item Name                                                  
178     Oathbreaker, Last Hope of the Breaking Storm        4.23   
145     Fiery Glass Crusader                                4.58   
108     Extraction, Quickblade Of Trembling Hands           3.53   
82      Nirvana                                             4.90   
19      Pursuit, Cudgel of Necromancy                       1.02   

                                                      Total_Purchase_Value  
Item ID Item Name                                                           
178     Oathbreaker, Last Hope of the Breaking Storm                 50.76  
145     Fiery Glass Crusader                                         41.22  
108     Extraction, Quickblade Of Trembling Hands                    31.77  
82      Nirvana                                                      44.10  
19      Pursuit, Cudgel of Necromancy                                 8.16

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [377]:
# Sort above DataFrame with Total Purchase Value
most_profitable_items_df = most_popular_items_df.sort_values("Total_Purchase_Value", ascending=False)
most_profitable_items_df.head(5)

# Formatting the DataFrame
most_popular_items_df["Item_Price"] = most_popular_items_df["Item_Price"].map("${:.2f}".format)
most_popular_items_df["Total_Purchase_Value"] = most_popular_items_df["Total_Purchase_Value"].map("${:.2f}".format)
most_popular_items_df.head(5)


Purchase_Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
108     Extraction, Quickblade Of Trembling Hands                  9   
82      Nirvana                                                    9   
19      Pursuit, Cudgel of Necromancy                              8   

                                                     Item_Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
108     Extraction, Quickblade Of Trembling Hands         $3.53   
82      Nirvana                                           $4.90   
19      Pursuit, Cudgel of Necromancy                     $1.02   

                                                     Total_Purchase_Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
108     Extraction, Quickblade Of Trembling Hands                  $31.77  
82      Nirvana                                                    $44.10  
19      Pursuit, Cudgel of Necromancy                               $8.16

CONCLUSIONS: Within this data and each table created, we can conclude many different things. This purchases within the game are mainly male dominated with a number of 652 individual players that made purchases. The next biggest demographic is female with only 113 players that made purchases. The age group that makes the most purchases within the game are those between the ages of 20 and 24 with the highest total purchase count and highest total purchase value. The players between the ages of 35 and 39 has the highest average purchase price and the highest total purchase per person overall. The item "Oathebreaker, Last Hope of the Breaking Storm" is both the most popular item and the most profitable item overall. In summary, the target demographic is males between the ages of 20 through 24. You could possibly market towards 35 and 39 year olds since they have the highest average purchse price and highest total value. If we could increase this demographic, we could potentially increase overall revenue for the company.